fix(parser): allocate all strings in arena

overlookmotel opened the pull request:
Fix #11723.

Fix 2 cases where parser was using static strings for AST nodes.

This is fine on Rust side, but it breaks raw transfer, because raw transfer requires all data (including strings) to be in the arena. All it knows is what's in the arena, and a `&'static str` is stored outside that, so it can't read the string data.

Unfortunately, we have no way to statically prevent putting an `Atom` created from a `&'static str` in AST, because type system allows a `'static` lifetime to be "squeezed" down to `'a`. So we have to play "whack-a-mole" with any such bugs at present (unfortunately conformance tests don't cover the cases in #11723).

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11738?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#11738** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11738?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11738?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#11737** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11737?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=11738).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Boshen commented:
### Merge activity

* **Jun 16, 11:13 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
fix(parser): allocate all strings in arena (#11738)

Fix #11723.

Fix 2 cases where parser was using static strings for AST nodes.

This is fine on Rust side, but it breaks raw transfer, because raw transfer requires all data (including strings) to be in the arena. All it knows is what's in the arena, and a `&'static str` is stored outside that, so it can't read the string data.

Unfortunately, we have no way to statically prevent putting an `Atom` created from a `&'static str` in AST, because type system allows a `'static` lifetime to be "squeezed" down to `'a`. So we have to play "whack-a-mole" with any such bugs at present (unfortunately conformance tests don't cover the cases in #11723).
